xcm-support/src/lib.rs

```diff
@@ -75,3 +78,79 @@ impl UnknownAsset for () {
 		Err(DispatchError::Other(NO_UNKNOWN_ASSET_IMPL))
 	}
 }
+
+/// Extracts the `AccountId32` from the passed `location` if the network
+/// matches.
+pub struct RelaychainAccountId32Aliases<Network, AccountId>(PhantomData<(Network, AccountId)>);
+impl<Network: Get<NetworkId>, AccountId: From<[u8; 32]> + Into<[u8; 32]> + Clone>
+	xcm_executor::traits::Convert<MultiLocation, AccountId> for RelaychainAccountId32Aliases<Network, AccountId>
+{
+	fn convert(location: MultiLocation) -> Result<AccountId, MultiLocation> {
+		let id = match location {
+			MultiLocation {
+				parents: 1,
+				interior: X1(AccountId32 {
+					id,
+					network: NetworkId::Any,
+				}),
+			} => id,
+			_ => return Err(location),
+		};
+		Ok(id.into())
+	}
+
+	fn reverse(who: AccountId) -> Result<MultiLocation, AccountId> {
+		Ok(AccountId32 {
+			id: who.into(),
+			network: Network::get(),
+		}
+		.into())
+	}
+}
+
+/// when Relay an XCM message from a given `interior` location, if the given
+/// `interior` is not `Here`, the destination will receive a xcm message
+/// beginning with `DescendOrigin` as the first instruction. so the xcm message
+/// format must match this order:
+/// `DescendOrigin`,`WithdrawAsset`,`BuyExecution`,`Transact`.
+pub struct AllowEquivalentAccountsFrom<T>(PhantomData<T>);
+impl<T: Contains<MultiLocation>> ShouldExecute for AllowEquivalentAccountsFrom<T> {
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		message: &mut Xcm<Call>,
+		max_weight: Weight,
+		_weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		ensure!(T::contains(origin), ());
+		let mut iter = message.0.iter_mut();
```

Comment:
Why does it have to be mutable?

Reply:
Because `Iterator::next` is a mutable method, and it's used in the line below.

Reply:
I meant the iterator (`iter_mut`). And I just found the mutation below `*weight = max_weight;` so never mind.